dmutils/request_id.py

```diff
@@ -1,50 +1,90 @@
 import uuid
+from itertools import chain
 
 from flask import request, current_app
-from flask.wrappers import Request
 
 
-class CustomRequest(Request):
-    _request_id = None
-
+class RequestIdRequestMixin(object):
     @property
     def request_id(self):
-        if self._request_id is None:
-            self._request_id = self._get_request_id(
-                current_app.config['DM_REQUEST_ID_HEADER'],
-                current_app.config['DM_DOWNSTREAM_REQUEST_ID_HEADER'])
-        return self._request_id
-
-    def _get_request_id(self, request_id_header, downstream_header):
-        if request_id_header in self.headers:
-            return self.headers.get(request_id_header)
-        elif downstream_header and downstream_header in self.headers:
-            return self.headers.get(downstream_header)
+        return self.trace_id
+
+    @property
+    def trace_id(self):
+        if not hasattr(self, "_trace_id"):
+            self._trace_id = self._get_first_header(current_app.config['DM_TRACE_ID_HEADERS']) or str(uuid.uuid4())
+        return self._trace_id
+
+    @property
+    def span_id(self):
+        if not hasattr(self, "_span_id"):
+            self._span_id = self._get_first_header(current_app.config['DM_SPAN_ID_HEADERS'])
+        return self._span_id
+
+    @property
+    def parent_span_id(self):
+        if not hasattr(self, "_parent_span_id"):
+            self._parent_span_id = self._get_first_header(current_app.config['DM_PARENT_SPAN_ID_HEADERS'])
+        return self._parent_span_id
+
+    def _get_first_header(self, header_names):
+        """
+        Returns value of request's first present (and Truthy) header from header_names
+        """
+        for header_name in header_names:
+            if header_name in self.headers and self.headers[header_name]:
+                return self.headers[header_name]
         else:
-            return str(uuid.uuid4())
+            return None
+
+    def get_onwards_request_headers(self):
```

Comment:
How is this getting used? From a few minutes of looking I can't see this being called anywhere?

Reply:
It isn't - yet. The idea is it *will* be checked by e.g. the apiclient. See the paragraph beginning "Added a centralized..." in my initial description.